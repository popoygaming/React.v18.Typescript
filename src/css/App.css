

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.app-body{
  align-items: center;
}