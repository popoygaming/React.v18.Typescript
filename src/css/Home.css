.main-container{
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 80%;
  min-width: 100px;
}
.books-container{
  display: flex;
  justify-content: space-between; 
  align-items: stretch;
}
.book-collection{
  width: 100%;
}
.space-between{
  width: 40px;
}
span {
  cursor: pointer;
  color: #1976d2;
  text-decoration: underline;
}

@media (max-width: 600px) {
  .books-container {
    flex-direction: column;
  }
  .space-between{
    height: 20px;
  }
}