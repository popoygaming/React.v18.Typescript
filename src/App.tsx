import "./css/App.css";
import Header from "./components/Header";
import { HEADER_TITLE } from "./utils/constants";
import { RouterProvider } from "react-router-dom";
import router from "./router/router";

function App() {
  return (
    <html lang="en">
      <header>
        <Header title={HEADER_TITLE}></Header>
      </header>
      <body className="app-body">
        <RouterProvider router={router} />
      </body>
    </html>
  );
}

export default App;
