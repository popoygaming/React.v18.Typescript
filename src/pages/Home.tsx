import React from "react";
import "../css/Home.css";
import { Button, Divider, Link, Stack, Typography } from "@mui/material";
import BooksCollection from "../components/BooksCollection";
import {
  BOOKS_FEMALE_TITLE,
  API_URL_BOOKS,
  BOOKS_MALE_TITLE,
  GET_BOOKS_TEXT,
  Gender,
  HANDCOVER_BOOKS_FEMALE_TITLE,
  HANDCOVER_BOOKS_MALE_TITLE,
  HARDCOVER_ONLY_TEXT,
} from "../utils/constants";
import useFetch from "../hooks/useFetch";
import BooksService from "../services/BookService";

function Home() {
  const [femaleBooks, setFemaleBooks] = React.useState<IBook[]>([]);
  const [maleBooks, setMaleBooks] = React.useState<IBook[]>([]);
  const [isHardcoverOnly, setHandcoverOnly] = React.useState(false);
  const [isGetBooksFocused, setGetBooksFocused] = React.useState(false);
  const [isHandcoverFocused, setHandcoverFocused] = React.useState(false);

  const getBooksButtonRef = React.useRef(document.createElement("div"));
  const handcoverOnlyLinknRef = React.useRef(document.createElement("div"));

  // const { retData, isPending, error } = useFetch(`${API_URL_BOOKS}/books`);

  const handleGetBooks = async () => {
    setHandcoverOnly(false);
    setFemaleBooks(await BooksService.get(Gender.Female, false));
    setMaleBooks(await BooksService.get(Gender.Male, false));
  };

  const handleHandcoverOnly = async () => {
    setHandcoverOnly(true);
    setFemaleBooks(await BooksService.get(Gender.Female, true));
    setMaleBooks(await BooksService.get(Gender.Male, true));
  };

  // accessiblity - keyboard functions
  const handleGetBooksBtnBlur = () => {
    setGetBooksFocused(false);
  };
  const handleGetBooksBtnFocus = () => {
    setGetBooksFocused(true);
  };
  const handleHandcoverBlur = () => {
    setHandcoverFocused(false);
  };
  const handlehandcoverFocus = () => {
    setGetBooksFocused(true);
  };
  const handleGetBooksKeyDown = (e: any) => {
    if (e.key === "Enter" || e.key === "") {
      e.preventDefault();
      getBooksButtonRef.current.click();
    }
  };
  const handleHandcoverOnlyKeyDown = (e: any) => {
    if (e.key === "Enter" || e.key === "") {
      e.preventDefault();
      handcoverOnlyLinknRef.current.click();
    }
  };

  return (
    <div className="main-container">
      <div className="books-container">
        <div className="book-collection">
          <BooksCollection title={ isHardcoverOnly ? HANDCOVER_BOOKS_MALE_TITLE : BOOKS_MALE_TITLE} books={maleBooks} data-testid="test-booksCollection"></BooksCollection>
        </div>
        <div className="space-between" />
        <div className="book-collection">
          <BooksCollection title={ isHardcoverOnly ? HANDCOVER_BOOKS_FEMALE_TITLE : BOOKS_FEMALE_TITLE } books={femaleBooks}></BooksCollection>
        </div>
      </div>
      <br />
      <Divider />
      <br />
      <Stack spacing={{ xs: 1, sm: 2 }} direction={"row"} useFlexGap flexWrap="wrap" alignItems={"center"} justifyContent={"right"}>
        <Link href="#" onClick={handleHandcoverOnly} data-testid="test-getHandcoverOnly" onFocus={handlehandcoverFocus} onBlur={handleHandcoverBlur} tabIndex={isGetBooksFocused ? 1 : 0} onKeyDown={handleHandcoverOnlyKeyDown}>
          <Typography variant="subtitle2" ref={handcoverOnlyLinknRef}>{HARDCOVER_ONLY_TEXT}</Typography>
        </Link>
        <Button variant="contained" data-testid="test-getBooks" onClick={handleGetBooks} onFocus={handleGetBooksBtnFocus} onBlur={handleGetBooksBtnBlur} tabIndex={isHandcoverFocused ? 1 : 0} onKeyDown={handleGetBooksKeyDown}>
          <Typography variant="subtitle2" ref={getBooksButtonRef}>{GET_BOOKS_TEXT}</Typography>
        </Button>
      </Stack>
    </div>
  );
}

export default Home;
