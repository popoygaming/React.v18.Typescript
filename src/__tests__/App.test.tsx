import React from 'react';
import { render, screen} from '@testing-library/react';
import App from '../App';
import Home from '../pages/Home';
import renderer from 'react-test-renderer';

it('performs snapshot testing', () => {
  // const tree = renderer.create(<Home />).toJSON();
  const tree = renderer.create(<Home />);
  expect(tree).toMatchSnapshot();
});