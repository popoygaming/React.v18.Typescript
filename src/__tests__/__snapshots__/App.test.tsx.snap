// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performs snapshot testing 1`] = `
<div
  className="main-container"
>
  <div
    className="books-container"
  >
    <div
      className="book-collection"
    >
      <div
        className="card"
      >
        <div
          className="card-header"
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            Books with Male Owners
          </h6>
        </div>
        <div
          className="card-content"
        >
          <ul
            className="book-list"
          >
            <h6
              className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
              data-testid="test-noBooks"
            >
              No books available.
            </h6>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="space-between"
    />
    <div
      className="book-collection"
    >
      <div
        className="card"
      >
        <div
          className="card-header"
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            Books with Female Owners
          </h6>
        </div>
        <div
          className="card-content"
        >
          <ul
            className="book-list"
          >
            <h6
              className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
              data-testid="test-noBooks"
            >
              No books available.
            </h6>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <br />
  <hr
    className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
  />
  <br />
  <div
    className="MuiStack-root css-w1eqp8-MuiStack-root"
  >
    <a
      className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1xylxj1-MuiTypography-root-MuiLink-root"
      data-testid="test-getHandcoverOnly"
      href="#"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
      >
        Hardcover Only
      </h6>
    </a>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="test-getBooks"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
      >
        Get Books
      </h6>
    </button>
  </div>
</div>
`;
